File: src/main/java/com/joraph/schema/Property.java
package com.joraph.schema;

import kotlin.jvm.functions.Function1;

/**
 * Implementers have reflection-based runtime access to a
 * class's properties.
 * @param <T> the property type
 */
@FunctionalInterface
public interface Property<T, R>
		extends Function1<T, R> {

	@Override
	default R invoke(T value) {
		return read(value);
	}

	R read(Object obj);

}

File: src/main/kotlin/com/joraph/schema/Property.kt
package com.joraph.schema


/**
 * Base interface representing a property of an entity with an access method.
 */
@FunctionalInterface
interface Property<T, R> : (T) -> R? {
    override fun invoke(value: T): R? {
        return read(value as Any)
    }

    fun read(obj: Any): R?
}
